hts.c

```diff
@@ -428,6 +428,9 @@ htsFile *hts_open_format(const char *fn, const char *mode, const htsFormat *fmt)
     if (strchr(mode, 'w') && fmt && fmt->compression == bgzf) {
         if (fmt->format == sam || fmt->format == vcf || fmt->format == text_format)
             *mode_c = 'z';
+    } else if (!*mode_c && !strchr(smode, 'z') &&
+               strlen(fn) > 3 && strcmp(fn+strlen(fn)-3, ".gz") == 0) {
+        *mode_c = 'z';
```

Comment:
This should be in `sam_open_mode()` instead. Samtools uses that function so adding "z" based on the filename in `sam_open_mode()` should also have the effect of permitting samtools to output compressed SAM.

Reply:
Thanks.  I'll take a look.

I confess I get a bit confused at times as to which function is the official blessed way to open files.  The htslib test_view program for example uses hts_open to read and write files, and this has been working with "z" for ages.

I'd be surprised if z wasn't permitted with sam_open already.  This change wasn't to add that, but to *automatically* add it if the filename ended in .gz.

Reply:
Perhaps it's slightly unfortunately named: `sam_open_mode()` doesn't open files!

Instead it constructs an “open mode” string based on the filename given to it, to be later given as the mode parameter to `hts_open`/`sam_open` of whatever flavour.

TODO it's got no doco blurb in _htslib/sam.h_…

Reply:
Updated `sam_open_mode`, `sam_open_mode_format` and "in for a penny, in for a pound" `hts_parse_format` too.

This means we can now do e.g. "samtools view -O sam.gz,level=1 foo.bam | whatever".  I'm not sure that's actually helpful, but it's consistent.